components/header.js

```diff
@@ -33,7 +33,8 @@ function WalletSummary ({ me }) {
   if (me.privates?.hideWalletBalance) {
     return <HiddenWalletSummary abbreviate fixedWidth />
   }
-  return `${abbrNum(me.privates?.sats)}`
+  const sats = me.privates?.sats + me.weblnSats
```

Comment:
if either of these are undefined, I think it will cause a NaN value. Also, does it even make sense to combine the two values? you can only spend from one or the other

Reply:
> if either of these are undefined, I think it will cause a NaN value.

You're right, it would cause a NaN value in that case. But I made sure in _components/me.js:15_ that `me.weblnSats` is never undefined. It's set to 0 in that case.

However, the optional chaining in `me.privates?.sats` makes this look like _it_ might be undefined, yes. I'll double-check that this can actually never be the case, thanks! I think we're just using optional chaining in some cases where it's not really needed.

> Also, does it even make sense to combine the two values? you can only spend from one or the other

Good point, but I think it does make sense. Currently, you can only spend from one or the other. If the user does not have enough sats in their custodial wallet, we show a QR invoice which then get paid from this "WebLN" balance (not really happy with the name).

But with some clever logic, I think we'll be able to spend from both wallets at the same time. For example, imagine someone only has 50 sats in their custodial wallet (`me.privates.sats`) and wants to pay a 100 sat invoice.

We should be able to use 50 sats from their custodial wallet and the remaining via WebLN (assuming they have enough sats there).

So I think it makes sense to keep these balances separate in the frontend code so we can implement such a thing by splitting invoices, for example.

Reply:
how would the combined balance work in practice? you'd need to send some sats to the custodial wallet and then pay the original invoice from the custodial wallet, right? I think this is probably too complicated and also might not even be needed (ideally users either go full self-custodial or full custodial)

Reply:
> I think this is probably too complicated and also might not even be needed (ideally users either go full self-custodial or full custodial)

Users can get zapped on SN. So until we also have implemented receiving self-custodial (which is going to be nontrivial since we need to be able to take a fee for ranking and sybil resistance - see discussion in #524), it's going to be natural that users will use both wallets at the same time in the mean time.

Also, we don't intend to drop the custodial wallet anytime soon. It's good for onboarding users.

So in the case users want to start with self-custody, it's highly likely they already have some sats on SN. What to do with these sats if they want to connect a self-custodial wallet?

Reply:
Until we have self-custodial receives or people stop wanting to receive money, we'll need to custody receives. 

One alternative to combining balances is to *only* display custodial balances on SN and surfacing the self-custodial balance elsewhere.

Reply:
Since Mutiny will probably not support `get_balance` due to valid privacy concerns (see [this comment](https://github.com/MutinyWallet/mutiny-node/issues/895#issuecomment-1851285461) in https://github.com/MutinyWallet/mutiny-node/issues/895) and since we already have a UI/UX problem with mixing sats from two wallets, I think it makes sense to not show the self-custodial balance at all on SN. Users can always look up their self-custodial balance themselves in the connected wallet.

Also, @rolznz does have a point with this that I didn't really acknowledge in my [previous reply](https://github.com/stackernews/stacker.news/pull/715#discussion_r1437614064). Mixing sats might indeed be more complicated than really necessary:

>  I think this is probably too complicated and also might not even be needed

So this means we would only show the custodial balance on SN (as before). The self-custodial balance is unknown to us but I believe if we implement zapping as follows, this shouldn't be a problem:

Assume we have X sats on SN and we have a WebLN provider connected.

If we now want to pay for something which costs Y sats with Y > X, we first ask the WebLN provider to pay for Y - X sats with a HODL invoice. If that succeeds, we know we can fully pay the invoice with the remaining known X sats. We then use these custodial sats to pay the remaining amount. If that also succeeds (which should be the case in 99.999% of all cases since it would basically use the same code as fully custodial zaps did all this time), we settle the HODL invoice. Payment done and we didn't need to know the self-custodial balance in advance. Using HODL invoices also shouldn't be a problem since all of this should only take a few seconds to settle. So from the PoV of the WebLN provider, it won't even notice it's a HODL invoice since it gets settled (or canceled) pretty quick.

Any thoughts on this @huumn @rolznz ?

Reply:
I think it might be simpler to not doing any sort of combined payments, and just allow users to withdraw sats from custody to their own wallet (that already works)

Is there any downside there?

Reply:
> (that already works)

With that, you mean that works for some WebLN providers? Or does it work for all of them already? I mentioned in the call yesterday that I wasn't thinking about Bitcoin Connect for receiving sats yet. So I might not be up to speed regarding this and may have missed some important parts about Bitcoin Connect. I only thought of it for sending sats so far.

> Is there any downside there?

Yes, there is. Even if they would withdraw their sats to their own wallet, they will still receive sats into their custodial wallet. So this problem does not go away with withdrawing sats before connecting.

But we don't even want people to withdraw their sats just to connect a wallet. That would be annoying. Maybe they just want to try out how connecting a wallet works without fully committing to it just yet? Maybe they don't feel comfortable with self-custody yet so they don't feel comfortable with withdrawing their whole balance from SN just yet? etc.

Reply:
> With that, you mean that works for some WebLN providers? Or does it work for all of them already? I mentioned in the call yesterday that I wasn't thinking about Bitcoin Connect for receiving sats yet. So I might not be up to speed regarding this and may have missed some important parts about Bitcoin Connect. I only thought of it for sending sats so far.

What I mean is stacker news already supports paying an invoice for a withdrawal (and it also supports WebLN.makeInvoice). That invoice can come from any wallet.

> So this problem does not entirely go away with withdrawing sats before connecting.
I don't mean before connecting. I just mean if the user wants to withdraw profits then they can withdraw to their own wallet.

I just think that trying to combine self-custodial + custodial to do payments (especially when partially paying from both) is going to be complicated. I think it would be easier if you connect a wallet using Bitcoin Connect, then zaps should come from your connected wallet. Otherwise use the custodial stacker news wallet. If they just want to test out self-custodial zaps and then go back to custodial zaps, they could just disconnect their wallet.

Reply:
> If we now want to pay for something which costs Y sats with Y > X, we first ask the WebLN provider to pay for Y - X sats with a HODL invoice.

I think this makes sense. If it's any easier, having WebLN pay Y directly is probably sufficient for MVP when Y > X. But your intuition is correct in that paying for Y as `Y - X non-custodial + X custodial` is the ideal. It's like paying at a store with half cash and half credit.

> I just think that trying to combine self-custodial + custodial to do payments (especially when partially paying from both) is going to be complicated.

Our plan is to eventually do that but we aren't ready to rely on an external wallet 100% yet. When we do that we inherit a user experience that we can't fully control and are inexperienced troubleshooting. Also, if a stacker has a balance on SN, paying with that balance has zero fees and paying with a non-custodial wallet has variable fees. Defaulting to the lower fee option seems like better UX. Further, our code is designed around this custodial model so supplementing it with an external wallet is going be easier to start.

We plan on doing something similar for non-custodial receives. Stackers will receive to their custodial wallets but we'll push out funds to an external wallet if their custodial wallet breaches a certain threshold.

In both cases, once we've nailed supplementing the custodial experience, we'll move to a model where there is no hybrid wallet.